### PR TITLE
JM paper remarks

### PR DESCRIPTION
**Pull request by jannes-m**

### CODE REVIEW

**Comment by jannes-m** [comment]
Hey Patrick,
I have read the complete paper, however, I could only review until the discussion section. I still hope it helps. In any case, it is great to see that this will be soon out there and it is so Patrick-like, really great stuff!!
All the best,
Jannes

File: paper/paper.Rmd [hunk 1]
```diff
@@ -606,6 +611,7 @@ Considering the wealth and continuing development of spatial and spatiotemporal
 At present, \pkg{mlr3spatiotempcv} also supports spatiotemporal partitioning using the versatile CLUTO clustering algorithm [@cluto].
 CLUTO is available in R through the \pkg{skmeans} package, which provides an interface to a downloadable compiled library with a restriction to non-commercial uses (see `?ResamplingSptCVCluto` in \pkg{mlr3spatiotempcv} for more information).
 With respect to functionality, \pkg{mlr3spatiotempcv} supports two-dimensional clustering based on coordinates and three-dimensional clustering on coordinates and time.
+<!--JM: not clear why feature clustering is not available when you already have presented the enviromental blocking based on specific features-->
 Feature clustering is not supported.
```
Comment (pat-s) [inline paper/paper.Rmd:615]: Good catch, I cannot remember if this is a leftover or why I did not spot this earlier!

Changed it to 

> With respect to functionality, \pkg{mlr3spatiotempcv} supports two-dimensional clustering based on coordinates, time or features and three-dimensional clustering on coordinates and time.

**Review by pat-s — COMMENTED** [review]
Thanks a ton, Jannes! The manuscript just got a lot better thanks to you.

I've replied inline where some replies might have been needed and will merge the changes I made in the `paper` branch directly into here and then finally merge this branch when all points are resolved.

File: paper/JMNotes.Rmd [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+- In case, you didn't know: from 1.0.0 onwards, sf uses s2 by default, i.e. spherical geometries.
+Hence, I am not sure if there is really much more work to do (see https://r-spatial.github.io/sf/articles/sf7.html).
+I am referring to this sentence in the paper:
+
+>It is planned to provide a more consistent interface that supports all CRS and allows the use of distance measures in projected planar space or using great circle distances on the unprojected ellipsoid, in the case of unprojected CRS.
+
+- The leave-one-disc-out method seems to be quite promising since the direct neighbors between the training and test sets are removed thus removing spatial autocorrelation which might still be present when using the kmeans clustering approach.
+In the back of my mind, I knew that this was possible but I didn't know that this was already implemented in **sperrorest**.
+I wonder why AB has often preferred the kmeans clustering approach over the leave-one-disc-out approach.
+Maybe something which can be discussed.
```
[the remark this answers is not in the document]
Comment (alexanderbrenning) [reply paper/JMNotes.Rmd:32]: LODO-CV:

* Yes, sperrorest does support LODO with and without a buffer.
* Remember you must fit a model on each training sample. When discs are small (extreme case: size 1 as in standard LOO), this means that a model is fitted for making only one prediction. This is computationally extremely expensive.
* When buffer size is large relative to the size of the study area, a large percentage of the study is not being sampled.
* Patrick, avoid claims of one method being better or less biased unless a clear concept of bias is associated with these claims. The questions are: 1. What do I want to estimate (i.e. what is appropriate for the application at hand. 2. How can I estimate it. - There is no bias in 2. as long as there's no answer to 1. I think I tried to clarify this in my background section, but I'll double-check this in the next pass. Overall, I think it is beyond the scope of the paper to make specific recommendations to use one or another method. We just don't know what people want to do in their analysis. The only thing mlr3 offers is an estimation procedure.
* This being said, I have something coming up that involves spLOO (and by extension, LODO) and comparison with kmeans...

File: paper/JMNotes.Rmd [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+- abstract: it is at least uncommon to cite other works in the abstract.
+If not explicitly required by the journal, I would refrain from it since you most likely cite all these works in the paper itself.
+- The last sentence of the abstract does not add much.
+Delete it or rewrite in a more compelling way.
+
+- I like the crop classification example, however, it is not entirely clear to me what agricultural districts are. 
+And though the crop classification introduces non-standard spatial resampling strategies, it does not make reference to the more common spatial resampling procedure (k-means clustering) where clustered coordinates are considered, which might be a more gentle introduction to spatial CV for readers who are not already familiar with the concepts.
+But I might be wrong.
+Looking at Table 1, the spatial blocking/tiles approach seems to be even more popular than the kmeans clustering approach.
```
[the remark this answers is not in the document]
Comment (alexanderbrenning) [reply paper/JMNotes.Rmd:11]: The crop classification is not a bad fit either. Yes, kmeans may cut across some fields, but hey, the study area is like 30 or 40 km in diameter, therefore kmeans-based CV measures a different level of spatial generalization or transferability. I can re-check my wording but I do think the idea with the agricultural districts / ... is a good one. Ideally you'd want to apply your classifier to adjacent areas, not only within the same region. Btw the crop classification has four crops, not a single one, and yes there will always a class called "other" - but that doesn't invalidate the use of this classifier. Its ability to make meaningful predictions in adjacent regions is exactly what we test with kmeans CV...
So in other words: Field-level CV won't tell you how well a classifier will perform in adjacent regions. kmeans CV won't tell you how well it fills in those gaps within your study area. Neither of them will tell you how well you can predict individual missing grid cells within a field. It's all about the purpose of the study! There's no single best, optimal, unbiased, whatever estimator of performance.
Let's not worry too much about popularity of methods - many people apply the chosen CV type for lack of better knowledge, without matching a specific error estimator to the question at hand.
There's one practical aspect about kmeans though that is particularly appealing: It adapts very well to irregular study region geometries. Rectangular tiles don't (which is why I implemented a heuristic in sperrorest that tries to rearrange tiny fragments of tiles).

[the remark this answers is not in the document]
Comment (jannes-m) [reply paper/JMNotes.Rmd:11]: As I said, I liked the crop example. And I wrongly believed that the *kmeans* approach is the most popular. Hence, leave it as is, and maybe explain what agricultural districts refer to. And maybe emphasize throughout the paper (i.e., repeat it again and again in abstract, intro and discussion) that there is no out-of-the-box, one-fits-it-all approach but that the approach depends on the dataset at hand and the prediction task/purpose.

File: paper/JMNotes.Rmd [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+- Table 1: Wow, not too many papers have so far used spatial CV.
+Is this a comprehensive list?
+Have you somewhere listed the references by chosen strategy?
+If so, put it in the Appendix and reference the Appendix in the caption.
```
[the remark this answers is not in the document]
Comment (jannes-m) [reply paper/JMNotes.Rmd:16]: I think the reviewers will ask for it but you can also wait for them asking before compiling the corresponding information -.)

File: paper/JMNotes.Rmd [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+Not entirely clear, what the footnotes refer to.
+These seem to be studies which have used the corresponding spatial CV approach but it is not all of them (e.g., spcv_block has been used by 28 papers but only 3 are referenced).
```
[the remark this answers is not in the document]
Comment (jannes-m) [reply paper/JMNotes.Rmd:18]: Yes, at least put it into the table caption to make clear that you could not cite all works in the table.

File: paper/JMNotes.Rmd [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+These seem to be studies which have used the corresponding spatial CV approach but it is not all of them (e.g., spcv_block has been used by 28 papers but only 3 are referenced). 
+Please clarify.
+
+- I really do like the visualizations, great job! And they help the reader a lot to better grasp the differences between the resampling strategies.
+
+- In case, you didn't know: from 1.0.0 onwards, sf uses s2 by default, i.e. spherical geometries.
+Hence, I am not sure if there is really much more work to do (see https://r-spatial.github.io/sf/articles/sf7.html).
+I am referring to this sentence in the paper:
+
+>It is planned to provide a more consistent interface that supports all CRS and allows the use of distance measures in projected planar space or using great circle distances on the unprojected ellipsoid, in the case of unprojected CRS.
```
[the remark this answers is not in the document]
Comment (jannes-m) [reply paper/JMNotes.Rmd:27]: s2 geometries, i.e., computations on the sphere, are now used by **sf** for geographic (i.e. unprojected) coordinates (e.g., CRS:4326) by default. But this also means that as long as you provide a projected CRS, computations are still done on the plane. This again would mean that there might be more work to do if you really intended to provide a consistend interface that supports all CRS. This is just for your information. In any case, you should rewrite the sentence if you use **sf** under the hood for all geometric operations by replacing `using great circle distances on the unprojected ellipsoid, in the case of unprojected CRS` with `using spherical distances in the case of unprojected CRS`.

File: paper/JMNotes.Rmd (hunk 2 above)
[the remark this answers is not in the document]
Comment (jannes-m) [reply paper/JMNotes.Rmd:32]: Excellent, thanks for the clarification!

File: paper/JMNotes.Rmd [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+- The leave-one-disc-out method seems to be quite promising since the direct neighbors between the training and test sets are removed thus removing spatial autocorrelation which might still be present when using the kmeans clustering approach.
+In the back of my mind, I knew that this was possible but I didn't know that this was already implemented in **sperrorest**.
+I wonder why AB has often preferred the kmeans clustering approach over the leave-one-disc-out approach.
+Maybe something which can be discussed.
+
+- Figure 7: spatial tiling. 
+It would be great if one could see the used rectangular blocks.
```
[the remark this answers is not in the document]
Comment (jannes-m) [reply paper/JMNotes.Rmd:35]: Ok.

File: paper/JMNotes.Rmd [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+- Figure 9: in fact, one only sees 4 time levels, not 5 as stated in the caption
```
[the remark this answers is not in the document]
Comment (jannes-m) [reply paper/JMNotes.Rmd:36]: Then please explain so in the caption otw other readers might also become confused.

File: paper/paper.Rmd [hunk 9]
```diff
@@ -328,12 +332,12 @@ Implementations of blocking methods often provide additional options to fine tun
...
-Supplying a numeric value in meters via argument `theRange` (named just `range` in {mlr3spatiotempcv}) will create a quadratic block pattern across the study area.
-Ultimately, users can also supply a user-defined polygon which defines the zones.
+The "Spatial blocking" from package \pkg{blockCV} comes with various blocking options such as random (default) (`selection = "random"`), systematic (`selection = "systematic"`) or checkerboard pattern (`selection = "checkerboard"`).
+Additionally, the user might supply a numeric value in meters via argument `theRange` (named just `range` in {mlr3spatiotempcv}) which will ensure a quadratic block pattern across the study area.
+<!--JM: to stay consistent please add the argument which lets the user indicate a user-defined polygon or are you referring to the rectangular block design specified with rows and cols? If so, say so. If not, clarify and additionally mention the rectangular block design here-->
```
[the remark this answers is not in the document]
Comment (jannes-m) [reply paper/paper.Rmd:337]: If you don't think that `rasterLayer` is useful to the user, omit it! If you mention it, pls also mention argument `rasterLayer` since you offer the arguments for all other options.

File: paper/JMNotes.Rmd [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+- LLTO: maybe you could expand a little bit on why observations need to be omitted.
+For readers new to the LLTO concept, the reasoning for this might not be entirely clear.
```
[the remark this answers is not in the document]
Comment (jannes-m) [reply paper/JMNotes.Rmd:38]: I found Hanna's [conceptual figure](https://github.com/mlr-org/mlr3spatiotempcv/files/4142419/cv.pdf) quite enlightening, and I think it also does a better job of showing why and which observations have to be omitted than Figure 11 (it's just too many observations). Here's my take on the explanation (not sure if its easier to understand):
`In order to make a prediction for a specific time stamp of a specific location (test set), one would have to omit both the observations of the same location as well a the observations from the same time stamp of other locations in the training set.`

[the remark this answers is not in the document]
Comment (alexanderbrenning) [reply paper/JMNotes.Rmd:38]: > In order to make a prediction for a specific time stamp of a specific location (test set), one would have to omit both the observations of the same location as well a the observations from the same time stamp of other locations in the training set.

Why "have to omit"? There's nothing you have to remove by default. If my prediction task is to predict a future outcome at a location based on that location's history, then it doesn't make sense to remove its history from the training set. So let's refrain from telling the reader what they should be doing - we only offer technological solutions that may or may not fit their prediction tasks. We could use phrases such as "this resampling design may be particularly useful in situations in which..." to convince the reader that what was implemented may actually be useful in some relevant situation...

[the remark this answers is not in the document]
Comment (jannes-m) [reply paper/JMNotes.Rmd:38]: Basically, I only wanted to describe what LLTO is doing but I see your point. Ok, then let's rephrase in a more neutral tone:
If there are multiple observations over time per location, and the modeling purpose is to reduce the bias in the predictions as potentially introduced by spatial as well as temporal autocorrelation, LLTO might be an appropriate choice. To do so, LLTO makes predictions for a specific time stamp of a specific location (test set) by omitting the observations of this specific location as well a the observations of this specific time stamp of other locations in the training set.

File: paper/JMNotes.Rmd (hunk 3 above)
[the remark this answers is not in the document]
Comment (pat-s) [reply paper/JMNotes.Rmd:11]: @alexanderbrenning 

Fully agree. The most important point is that the splitting structure of the CV fits the application goal/purpose of the model. In this case the goal of the model would be to predict on the crop level, hence the CV should also happen on it.

@jannes-m 
Yes, we should probably make that point more often throughout the paper.

File: paper/JMNotes.Rmd (hunk 4 above)
[the remark this answers is not in the document]
Comment (pat-s) [reply paper/JMNotes.Rmd:16]: This is similar bad to the task of creating a table with all the 90 vegetation indices used :/

Not sure what the best format for this listing would be. Let's see :)

File: paper/JMNotes.Rmd (hunk 6 above)
[the remark this answers is not in the document]
Comment (pat-s) [reply paper/JMNotes.Rmd:27]: Thanks!

I am not sure if I want/can provide a consistent interface here given all the different underlying approaches and the substantial impact such a change would bring.
I really fear that users might get an unwanted result because they are not aware of this option/what it does and that it might come with more issues than it actually helps.

For now I've removed the sentence to avoid false promises.

File: paper/JMNotes.Rmd (hunk 2 above)
[the remark this answers is not in the document]
Comment (pat-s) [reply paper/JMNotes.Rmd:32]: Thanks Alex.

> Patrick, avoid claims of one method being better or less biased unless a clear concept of bias is associated with these claims.
> Overall, I think it is beyond the scope of the paper to make specific recommendations to use one or another method. 

Yes, I am trying to avoid that. Please highlight parts in which it might read like this.
It is not always easy to avoid that feeling with certain wordings but this is really an important point to get right here.

> This being said, I have something coming up that involves spLOO (and by extension, LODO) and comparison with kmeans...

Looking forward!

---


@alexanderbrenning 

A very specific point that would be of interest for me here (and maybe this is also partly what Jannes was referring to):

Why not choose `spcv_disc` with a "not too large" buffer (TBD how to determine the buffer size) over `spcv_coords` in scenarios where both would be appropriate? I.e. in an irregularly shaped study area where we would like to use a 5-10 fold CV? With a somewhat small buffer one should be able to reduce SAC at least a bit between train and test.

In most of your past studies you went with `spcv_coords` so it would be interesting to hear your thoughts about this :)

File: paper/JMNotes.Rmd (hunk 10 above)
[the remark this answers is not in the document]
Comment (pat-s) [reply paper/JMNotes.Rmd:38]: > I found Hanna's conceptual figure quite enlightening, and I think it also does a better job of showing why and which observations have to be omitted than Figure 11 (it's just too many observations).

For me the figure never easily clicked in my head but I think it can be quite helpful for others!

> So let's refrain from telling the reader what they should be doing - we only offer technological solutions that may or may not fit their prediction tasks. We could use phrases such as "this resampling design may be particularly useful in situations in which..." 

Yes, wording is really the key issue in this paper - similar to the other comment above where we discussed about how to "get it right for the reader" to avoid making false promises or strong statements.

---

Here is the new paragraph

> The "Leave-Location-and-Time-Out" (LLTO) method combines both "Leave-Time-Out" (LTO) and "Leave-Location-Out" (LLO).
It creates partitions that operate on spatiotemporal clusters to reduce both spatial and temporal autocorrelation.
If there are multiple observations over time per location, and the modeling purpose is to reduce the bias in the predictions as potentially introduced by spatial as well as temporal autocorrelation, LLTO might be an appropriate choice.
To reduce autocorrelation in both dimensions, the methods omits observations as the splitting in test and train sets only provides separation in one dimension whereas two (space and time) are needed to account for both types, spatial and temporal autocorrelation.
In other words, accounting for one dimension as in Fig. \@ref(fig:lto) or Fig. \@ref(fig:llo), by only having one time step or location in either test or training set, would have no effect on the respective other.
Thus, the removal of observation can be seen as a proxied buffering in multiple dimensions as shown in Fig. \@ref(fig:llto).

File: paper/paper.Rmd [hunk 11]
```diff
@@ -252,17 +252,19 @@ The available methods are further discussed in section \@ref(sec:disc).
...
 
 ### Buffering
 
 "Buffering" refers to the idea of removing observations between test and training set by drawing a circular buffer around one or more observations [@rest2014, @brenning2012, @blockCV].
 Within this buffer zone, all observations are removed from the training set, thereby reducing the similarity between training and test set.
 
-Buffering is mostly applied in the coordinate space due to the scattered appearance of observations in this space.
-However, there is no practical limitation that would prevent this method to be applied in a 2D + time setting.
-Because \pkg{mlr3spatiotempcv} supports two spatial-only "buffering" implementations, (1) "Spatial buffering" from package \pkg{blockCV} and (2) "Leave-one-disc-out cross-validation" from \pkg{sperrorest}, buffering was attributed to the category of spatial partitioning methods. 
+<!--JM: not entirely sure what scattered appearance refers to-->
```
[the remark this answers is not in the document]
Comment (pat-s) [reply paper/paper.Rmd:263]: Yes seems legit 👍

File: paper/paper.Rmd (hunk 9 above)
[the remark this answers is not in the document]
Comment (pat-s) [reply paper/paper.Rmd:337]: The underlying issue here is that I cannot/don't want to talk about every single parameter of the underlying method (in detail).
To me discussing the `selection` argument was more important than showcasing that one could also pass a custom polygon and then this one is used for the blocks.

I now mentioned the argument name in the sentence like

> Ultimately, users can also supply a user-defined polygon via argument `rasterLayer` to define the zones.

Do you think that's ok?

File: paper/JMNotes.Rmd (hunk 10 above)
[the remark this answers is not in the document]
Comment (jannes-m) [reply paper/JMNotes.Rmd:38]: IMHO still hard to follow but that might be just personal taste, so leave it as is, and wait for potential reviewer comments.

File: paper/paper.Rmd (hunk 9 above)
[the remark this answers is not in the document]
Comment (jannes-m) [reply paper/paper.Rmd]: I think I would emphasize the advantages of the mlr3 ecosystem in section **Should new methods be added to existing ML ecosystems or made available in a new package?**, rename the section accordingly, and also shortly refer to all the additional nice things this ecosystem offers. And in this section you could also name a few disadvantages.

[the remark this answers is not in the document]
Comment (jannes-m) [reply paper/paper.Rmd]: Now I need to go to sleep -:), the best of luck with your submission, JSS was at least famous for its overly long publication times (see [here](https://www.r-bloggers.com/2014/06/time-to-accept-it-publishing-in-the-journal-of-statistical-software/)) but maybe that has changed in the meantime.

[the remark this answers is not in the document]
Comment (pat-s) [reply paper/paper.Rmd]: Thanks! I renamed the section and rephrased a bit.

I am aware of the long review time in JSS but (luckily) this is not a blocker for the overarching plans :)

**Comment by pat-s** [comment]
@jannes-m Thanks again for your review and time! 🤝 
I'll merge the changes now into the main manuscript.
